### PR TITLE
Recover properly from SSM failures.

### PR DESCRIPTION
**Pull request by mattcollier**
@dlongley I'd like to get you to review this overall approach to dealing with keys in a transaction way.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -65,28 +65,76 @@ exports.runOperation = async ({operation}) => {
...
+  2. key pending key record is created in bedrock-kms
+  ssm request to generate key succeeds
+  <failure> kms does not record successful key generation (remove pending flag)
+  state:
+    kms -> pending key
+    ssm -> proper key
+
+  3. key pending key record is created in bedrock-kms
+  ssm request to generate key fails
+  key record is removed from bedrock-kms
```
Comment (mattcollier) [inline lib/index.js]: Presently, kms key records are not being removed, they are just left as pending.  I think this is fine.

**Comment by mattcollier** [comment]
@dlongley bump, there are more changes required in this module, I want to avoid a monolith PR.

File: lib/index.js [hunk 2]
```diff
@@ -65,28 +65,76 @@ exports.runOperation = async ({operation}) => {
...
+      'Key not found.',
+      'NotFoundError',
+      {key: keyId, httpStatusCode: 404, public: true});
+  }
 
-    // check for duplicate key
-    if(record) {
+  // if there is an existing key record, controller must match in all cases
+  if(record) {
+    if(record.key.controller !== controller) {
```
Comment (dlongley) [inline lib/index.js]: Could be `if(record && record.key.controller !== controller) {`

File: lib/index.js [hunk 3]
```diff
@@ -65,28 +65,76 @@ exports.runOperation = async ({operation}) => {
...
+      throw new BedrockError(
+        'Permission denied.',
+        // FIXME: this was a NotFoundError in the case of non-generation
+        // operations, was that important, possible better security?
```
Comment (dlongley) [inline lib/index.js]: Yeah, this was not found for that purpose ... not sure if other mechanisms could just get around that though. If not, we should put it back, but if it wasn't helpful, we can leave it as permission denied.

Comment (dlongley) [reply lib/index.js]: Also, if it we change it to permission denied, the status code should be 400 not 404.

File: lib/index.js (hunk 1 above)
Comment (dlongley) [reply lib/index.js]: Yeah, I can accept that.

File: lib/index.js (hunk 3 above)
Comment (mattcollier) [reply lib/index.js]: Seems like a timing analysis would easily reveal this deception.  I think we go with permission denied, which is the simplest implementation unless there's an objection.

**Pull request merged**